xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -0,0 +1,128 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+
+define(["jquery"], function ($) {
+
+  /**
+   * Map the element to its data object
+   * So that each instance of the livedata on the page handle there own data
+   */
+  var instancesMap = new WeakMap();
+
+
+
+  /**
+   * The init function of the logic script
+   * For each livedata element on the page, returns its corresponding data / API
+   * If the data does not exists yet, create it from the element
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata component
+   */
+  var init = function (element) {
+
+    if (!instancesMap.has(element)) {
+      // create a new logic object associated to the element
+      var logic = new Logic(element);
+      instancesMap.set(element, logic);
+
+      logic.changeLayout();
+    }
+
+    return instancesMap.get(element);
+  };
+
+
+  /**
+   * Class for a logic element
+   * Contains the Livedata data object and methods to mutate it
+   * Can be used in the layouts to display the data, and call its API
+   * @param {HTMLElement} element The HTML Element corresponding to the Livedata
+   */
+  var Logic = function (element) {
+    this.element = element;
+    this.data = JSON.parse(element.getAttribute("data-data") || "{}");
+    element.removeAttribute("data-data");
+    this.layouts = {};
+  };
+
+
+  /**
+   * Change layout to specified one, or default one if none specified
+   * @param {String} layoutId The id of the layout to load with requireJS
+   * @returns {Object} A jquery promise
+   */
+  Logic.prototype.changeLayout = function (layoutId) {
+    var self = this;
+    var defer = $.Deferred();
+
+    layoutId = layoutId || this.data.meta.defaultLayout;
+    if (layoutId === this.data.query.currentLayout) { return defer.reject(); }
+    if (!this.data.meta.layoutDescriptors[layoutId]) { return defer.reject(); }
+    if (this.data.meta.layouts.indexOf(layoutId) === -1) { return defer.reject(); }
+
+    // load layout based on it's filename
+    require([BASE_PATH + "layouts/" + this.data.meta.layoutDescriptors[layoutId].file],
+      // load success
+      function (Layout) {
+        // remove current layout from the page
+        if (self.data.query.currentLayout && self.layouts[self.data.query.currentLayout]) {
+          self.element.removeChild(self.layouts[self.data.query.currentLayout]);
+        }
+        // add layout element in loaded layouts list if not already loaded on the page
+        if (!self.layouts[layoutId]) {
+          self.layouts[layoutId] = new Layout(self.element);
+        }
+        // add new layout to the page
+        self.element.appendChild(self.layouts[layoutId]);
```

Comment:
Yes, that was an issue I thought about.
I first tried to return the whole layout object, but that would have implied that every layout should follow the same architecture, with an `element` property for instance containing the layout root element to be used in the Logic script. And also, the Logic script doesn't have to know much more about the layout than its HTML element. So that's why I decided to return only the layout element instead of the whole object. 

I know this is not a very common pattern, but that was the easiest way to generate a new HTML object through a class. I can change it to a simple function like the one you proposed if you prefer.

Reply:
Returning the DOM element is good. I would drop the ``new`` operator.